source/slang/slang-ir-check-recursion.cpp

```diff
@@ -123,4 +123,70 @@ void checkForRecursiveFunctions(TargetRequest* target, IRModule* module, Diagnos
     }
 }
 
+void checkForOutOfBoundAccess(IRModule* module, DiagnosticSink* sink)
```

Comment:
Should move this to its own file, `slang-check-out-of-bound-access.h/cpp`.

You should use the `InstPassBase` helper class to help you visit all child insts, instead of doing a hard-coded multi-level loop on the IRModule. See `slang-ir-simplify-for-emit.cpp` for an example of how to write an inst pass.